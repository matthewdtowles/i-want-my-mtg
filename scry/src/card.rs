pub mod ingestion_service;

mod mapper;
mod models;
mod repository;

pub use ingestion_service::CardIngestionService;
pub use models::Card;