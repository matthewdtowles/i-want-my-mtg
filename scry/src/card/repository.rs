use crate::{card::Card, database::ConnectionPool};
use anyhow::Result;
use sqlx::QueryBuilder;
use std::sync::Arc;
use tracing::{debug, error, info, warn};

#[derive(Clone)]
pub struct CardRepository {
    db: Arc<ConnectionPool>,
}

impl CardRepository {
    pub fn new(db: Arc<ConnectionPool>) -> Self {
        Self { db }
    }

    pub async fn save(&self, cards: &[Card]) -> Result<u64> {
        if cards.is_empty() {
            warn!("0 cards given, 0 cards saved.");
            return Ok(0);
        }
        let card_count = self.save_cards(cards).await?;
        let legality_count = self.save_legalities(cards).await?;
        info!(
            "Saved {} cards and {} legalities",
            card_count, legality_count
        );
        Ok(card_count)
    }

    pub async fn delete_all(&self) -> Result<u64> {
        self.delete_table(String::from("legality")).await?;
        let cards_deleted = self.delete_table(String::from("card")).await?;
        info!("{} cards deleted.", cards_deleted);
        Ok(cards_deleted)
    }

    async fn delete_table(&self, table: String) -> Result<u64> {
        let qb = QueryBuilder::new(format!("DELETE FROM {} CASCADE", table));
        let total_deleted = self.db.execute_query_builder(qb).await?;
        info!("{} {} entities deleted.", total_deleted, table);
        Ok(total_deleted)
    }

    async fn save_cards(&self, cards: &[Card]) -> Result<u64> {
        if cards.is_empty() {
            warn!("0 cards given, 0 cards saved.");
            return Ok(0);
        }
        debug!("Saving {} cards", cards.len());
        // TODO: evaluate: can we encapsulate INSERT INTO <table> (<model attrs.key>) (<model attrs.value>)
        let mut query_builder = QueryBuilder::new(
            "INSERT INTO card (
                    id, artist, has_foil, has_non_foil, img_src, 
                    is_reserved, mana_cost, name, number, oracle_text,
                    rarity, set_code, type
                )",
        );
        query_builder.push_values(cards, |mut b, card| {
            b.push_bind(&card.id)
                .push_bind(&card.artist)
                .push_bind(&card.has_foil)
                .push_bind(&card.has_non_foil)
                .push_bind(&card.img_src)
                .push_bind(&card.is_reserved)
                .push_bind(&card.mana_cost)
                .push_bind(&card.name)
                .push_bind(&card.number)
                .push_bind(&card.oracle_text)
                .push_bind(&card.rarity)
                .push_bind(&card.set_code)
                .push_bind(&card.type_line);
        });
        // TODO: evaluate if this can be encapsulated for reuse
        query_builder.push(
            " ON CONFLICT (id) DO UPDATE SET
            artist = EXCLUDED.artist,
            has_foil = EXCLUDED.has_foil,
            has_non_foil = EXCLUDED.has_non_foil,
            img_src = EXCLUDED.img_src,
            is_reserved = EXCLUDED.is_reserved,
            mana_cost = EXCLUDED.mana_cost,
            name = EXCLUDED.name,
            number = EXCLUDED.number,
            oracle_text = EXCLUDED.oracle_text,
            rarity = EXCLUDED.rarity,
            set_code = EXCLUDED.set_code,
            type = EXCLUDED.type
        WHERE
            card.artist IS DISTINCT FROM EXCLUDED.artist OR
            card.has_foil IS DISTINCT FROM EXCLUDED.has_foil OR
            card.has_non_foil IS DISTINCT FROM EXCLUDED.has_non_foil OR
            card.img_src IS DISTINCT FROM EXCLUDED.img_src OR
            card.is_reserved IS DISTINCT FROM EXCLUDED.is_reserved OR
            card.mana_cost IS DISTINCT FROM EXCLUDED.mana_cost OR
            card.name IS DISTINCT FROM EXCLUDED.name OR
            card.number IS DISTINCT FROM EXCLUDED.number OR
            card.oracle_text IS DISTINCT FROM EXCLUDED.oracle_text OR
            card.rarity IS DISTINCT FROM EXCLUDED.rarity OR
            card.set_code IS DISTINCT FROM EXCLUDED.set_code OR
            card.type IS DISTINCT FROM EXCLUDED.type",
        );
        match self.db.execute_query_builder(query_builder).await {
            Ok(rows_affected) => Ok(rows_affected),
            Err(e) => {
                error!("Failed to save {} cards: {}", cards.len(), e);
                Err(e)
            }
        }
    }

    async fn save_legalities(&self, cards: &[Card]) -> Result<u64> {
        let all_legalities: Vec<_> = cards.iter().flat_map(|c| &c.legalities).collect();

        if all_legalities.is_empty() {
            // Delete all legalities for these cards since they have none
            let card_ids: Vec<String> = cards.iter().map(|c| c.id.clone()).collect();
            if !card_ids.is_empty() {
                let mut delete_query =
                    QueryBuilder::new("DELETE FROM legality WHERE card_id = ANY(");
                delete_query.push_bind(card_ids);
                delete_query.push(")");
                self.db.execute_query_builder(delete_query).await?;
            }
            return Ok(0);
        }

        // Delete all existing legalities for these cards
        let card_ids: Vec<String> = cards.iter().map(|c| c.id.clone()).collect();
        let mut delete_query = QueryBuilder::new("DELETE FROM legality WHERE card_id = ANY(");
        delete_query.push_bind(card_ids);
        delete_query.push(")");
        self.db.execute_query_builder(delete_query).await?;

        // Insert all new legalities
        let mut query_builder = QueryBuilder::new("INSERT INTO legality (card_id, format, status)");
        query_builder.push_values(&all_legalities, |mut b, legality| {
            b.push_bind(&legality.card_id)
                .push_bind(&legality.format)
                .push_bind(&legality.status);
        });

        self.db.execute_query_builder(query_builder).await
    }
}
