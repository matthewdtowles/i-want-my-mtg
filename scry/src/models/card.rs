use serde::{Deserialize, Serialize};
use sqlx::FromRow;

use crate::models::CardRarity;

#[derive(Debug, FromRow, Serialize, Deserialize)]
pub struct Card {
    pub id: String,
    pub artist: Option<String>,
    pub has_foil: bool,
    pub has_non_foil: bool,
    pub img_src: String,
    pub is_reserved: bool,
    pub mana_cost: Option<String>,
    pub name: String,
    pub oracle_text: Option<String>,
    pub rarity: CardRarity,
    pub set_code: String,
    pub type_line: String,
    pub order: Option<i32>,
}
