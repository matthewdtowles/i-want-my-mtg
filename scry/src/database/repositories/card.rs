// src/database/repositories/card.rs
use anyhow::Result;
use sqlx::QueryBuilder;

use crate::database::DatabaseService;
use crate::models::Card;

pub struct CardRepository<'db> {
    db: &'db DatabaseService,
}

impl<'db> CardRepository<'db> {
    pub fn new(db: &'db DatabaseService) -> Self {
        Self { db }
    }

    pub async fn bulk_insert(&self, cards: &[Card]) -> Result<u64> {
        if cards.is_empty() {
            return Ok(0);
        }

        let mut query_builder = QueryBuilder::new(
            "INSERT INTO card (scryfall_id, name, mana_cost, type_line, oracle_text, set_code) ",
        );

        query_builder.push_values(cards, |mut b, card| {
            b.push_bind(&card.id)
                .push_bind(&card.name)
                .push_bind(&card.mana_cost)
                .push_bind(&card.type_line)
                .push_bind(&card.oracle_text)
                .push_bind(&card.set_code);
        });

        query_builder.push(
            " ON CONFLICT (scryfall_id) DO UPDATE SET 
            name = EXCLUDED.name,
            mana_cost = EXCLUDED.mana_cost,
            type_line = EXCLUDED.type_line,
            oracle_text = EXCLUDED.oracle_text,
            set_code = EXCLUDED.set_code
        ",
        );

        self.db.execute_query_builder(query_builder).await
    }

    pub async fn count(&self) -> Result<i64> {
        let query = "SELECT COUNT(*) FROM card";
        self.db.count(query).await
    }

    pub async fn find_by_set(&self, set_code: &str) -> Result<Vec<Card>> {
        let query = "SELECT id, scryfall_id, name, mana_cost, type_line, oracle_text, set_code, created_at, updated_at 
                     FROM card 
                     WHERE set_code = $1";

        self.db.fetch_all_with_param::<Card, &str>(query, set_code).await
    }

    pub async fn find_by_scryfall_id(&self, scryfall_id: &str) -> Result<Option<Card>> {
        let query = "SELECT id, scryfall_id, name, mana_cost, type_line, oracle_text, set_code, created_at, updated_at 
                     FROM card 
                     WHERE scryfall_id = $1";

        self.db.fetch_optional_with_param::<Card, &str>(query, scryfall_id).await
    }

    pub async fn search_by_name(&self, name: &str) -> Result<Vec<Card>> {
        let query = "SELECT id, scryfall_id, name, mana_cost, type_line, oracle_text, set_code, created_at, updated_at 
                     FROM card 
                     WHERE name ILIKE $1";

        let search_pattern = format!("%{}%", name);
        self.db.fetch_all_with_param::<Card, String>(query, search_pattern).await
    }
}
