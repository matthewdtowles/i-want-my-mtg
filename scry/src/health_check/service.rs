use crate::database::ConnectionPool;
use crate::health_check::models::{BasicHealthStatus, DetailedHealthStatus};
use anyhow::Result;
use std::sync::Arc;
use tracing::info;

pub struct HealthCheckService {
    connection_pool: Arc<ConnectionPool>,
}

impl HealthCheckService {
    pub fn new(connection_pool: Arc<ConnectionPool>) -> Self {
        Self { connection_pool }
    }

    pub async fn basic_check(&self) -> Result<BasicHealthStatus> {
        info!("Performing basic health check");

        let card_count = self.count_cards().await?;
        let price_count = self.count_prices().await?;
        let set_count = self.count_sets().await?;

        Ok(BasicHealthStatus {
            card_count,
            price_count,
            set_count,
        })
    }

    pub async fn detailed_check(&self) -> Result<DetailedHealthStatus> {
        info!("Performing detailed health check");

        let basic = self.basic_check().await?;
        let cards_with_prices = self.count_cards_with_prices().await?;
        let cards_without_prices = basic.card_count - cards_with_prices;

        Ok(DetailedHealthStatus {
            basic,
            cards_with_prices,
            cards_without_prices,
        })
    }

    async fn count_cards(&self) -> Result<i64> {
        self.connection_pool
            .count("SELECT COUNT(*) FROM card")
            .await
    }

    async fn count_prices(&self) -> Result<i64> {
        self.connection_pool
            .count("SELECT COUNT(*) FROM price")
            .await
    }

    async fn count_sets(&self) -> Result<i64> {
        self.connection_pool.count("SELECT COUNT(*) FROM set").await
    }

    async fn count_cards_with_prices(&self) -> Result<i64> {
        self.connection_pool
            .count("SELECT COUNT(DISTINCT card_id) FROM price")
            .await
    }
}
