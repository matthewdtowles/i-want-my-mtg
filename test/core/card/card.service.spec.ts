import { Test, TestingModule } from '@nestjs/testing';
import { Card } from '../../../src/core/card/card.entity';
import { CardService } from '../../../src/core/card/card.service';
import { CardDto } from '../../../src/core/card/dto/card.dto';
import { CardRepositoryPort } from '../../../src/core/card/ports/card.repository.port';
import { IngestionServicePort } from '../../../src/core/ingestion/ingestion.service.port';
import { TestUtils } from '../../test-utils';


describe('CardService', () => {
    const testUtils: TestUtils = new TestUtils();
    const mockSetNumber = 1;

    const mockSetCards: Card[] = testUtils.getMockSetCards(testUtils.MOCK_SET_CODE);
    const mockCardsWithName: Card[] = testUtils.getMockCardsWithName(3);
    const mockSavedCard: Card = testUtils.getMockCard(mockSetNumber, testUtils.MOCK_SET_CODE);
    const inputCard: Card = testUtils.getMockInputCard(mockSetNumber, testUtils.MOCK_SET_CODE);

    let service: CardService;
    let repository: CardRepositoryPort;
    let ingestionSvc: IngestionServicePort;

    const mockCardRepository: CardRepositoryPort = {
        save: jest.fn().mockResolvedValue(mockSetCards),
        findAllInSet: jest.fn().mockResolvedValue(mockSetCards),
        findAllWithName: jest.fn().mockResolvedValue(mockCardsWithName),
        findById: jest.fn().mockResolvedValue(mockSavedCard),
        findBySetCodeAndNumber: jest.fn().mockResolvedValue(mockSavedCard),
        findByUuid: jest.fn().mockResolvedValue(mockSavedCard),
        delete: jest.fn(),
    };

    const mockCardIngestion: IngestionServicePort = {
        fetchSetCards: jest.fn().mockResolvedValue(mockSetCards),
        fetchAllSetsMeta: jest.fn(),
        fetchSetByCode: jest.fn(),
    };

    beforeEach(async () => {
        const module: TestingModule = await Test.createTestingModule({
            providers: [
                CardService,
                {
                    provide: CardRepositoryPort,
                    useValue: mockCardRepository,
                },
                {
                    provide: IngestionServicePort,
                    useValue: mockCardIngestion,
                },
            ],
        }).compile();

        service = module.get<CardService>(CardService);
        repository = module.get<CardRepositoryPort>(CardRepositoryPort);
        ingestionSvc = module.get<IngestionServicePort>(IngestionServicePort);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should be defined', () => {
        expect(service).toBeDefined();
    });

    it('saves cards and returns saved cards', async () => {
        const repoSaveCard = jest.spyOn(repository, 'save');
        const createdCards: CardDto[] = await service.save(mockSetCards);
        expect(repoSaveCard).toHaveBeenCalledWith(mockSetCards);
        expect(createdCards).toEqual(mockSetCards);
    });

    it('finds all cards in given set by setCode', async () => {
        const repoFindAllInSet = jest.spyOn(repository, 'findAllInSet');
        const ingestFetchSetCards = jest.spyOn(ingestionSvc, 'fetchSetCards');
        const foundCards: CardDto[] = await service.findAllInSet(testUtils.MOCK_SET_CODE);
        expect(repoFindAllInSet).toHaveBeenCalledWith(testUtils.MOCK_SET_CODE)
        expect(ingestFetchSetCards).not.toHaveBeenCalled();
        expect(foundCards).toEqual(mockSetCards);
    });

    it('finds all cards in given set by setCode after ingesting set cards', async () => {
        const repoFindAllInSet = jest.spyOn(repository, 'findAllInSet');
        const ingestFetchSetCards = jest.spyOn(ingestionSvc, 'fetchSetCards');
        const repoSaveCards = jest.spyOn(repository, 'save');
        jest.spyOn(repository, 'findAllInSet').mockReturnValueOnce(Promise.resolve([]));
        const foundCards: CardDto[] = await service.findAllInSet(testUtils.MOCK_SET_CODE);
        expect(repoFindAllInSet).toHaveBeenCalledWith(testUtils.MOCK_SET_CODE);
        expect(ingestFetchSetCards).toHaveBeenCalledWith(testUtils.MOCK_SET_CODE);
        expect(repoSaveCards).toHaveBeenCalledWith(mockSetCards);
        expect(foundCards).toEqual(mockSetCards);
    });

    it('finds every instance of a card with given name', async () => {
        const repoFindAllWithName = jest.spyOn(repository, 'findAllWithName');
        const foundCards: CardDto[] = await service.findAllWithName(testUtils.MOCK_CARD_NAME);
        expect(repoFindAllWithName).toHaveBeenCalledWith(testUtils.MOCK_CARD_NAME);
        expect(foundCards).toEqual(mockCardsWithName);
    });

    it('finds a unique instance of a card by id', async () => {
        const repoFindById = jest.spyOn(repository, 'findById');
        const foundCard: CardDto | null = await service.findById(mockSetNumber);
        expect(repoFindById).toHaveBeenCalledWith(mockSetNumber);
        expect(foundCard).toEqual(mockSavedCard);
    });

    it('returns unique instance of a card by setCode and card number in that set', async () => {
        const repoFindBySetCodeAndNumber = jest.spyOn(repository, 'findBySetCodeAndNumber');
        const ingestFetchSetCards = jest.spyOn(ingestionSvc, 'fetchSetCards');
        const foundCard: CardDto = await service.findBySetCodeAndNumber(testUtils.MOCK_SET_CODE, mockSetNumber);
        expect(repoFindBySetCodeAndNumber).toHaveBeenCalledWith(testUtils.MOCK_SET_CODE, mockSetNumber);
        expect(ingestFetchSetCards).not.toHaveBeenCalled();
        expect(foundCard).toEqual(mockSavedCard);
    });

    it('returns unique instance of a card by setCode and card number in that set after ingesting set card', async () => {
        const repoFindBySetCodeAndNumber = jest.spyOn(repository, 'findBySetCodeAndNumber');
        const ingestFetchSetCards = jest.spyOn(ingestionSvc, 'fetchSetCards');
        const repoSaveCards = jest.spyOn(repository, 'save');
        jest.spyOn(repository, 'findBySetCodeAndNumber').mockReturnValueOnce(Promise.resolve(null));
        const foundCard: CardDto = await service.findBySetCodeAndNumber(testUtils.MOCK_SET_CODE, mockSetNumber);
        expect(repoFindBySetCodeAndNumber).toHaveBeenCalledWith(testUtils.MOCK_SET_CODE, mockSetNumber);
        expect(ingestFetchSetCards).toHaveBeenCalledWith(testUtils.MOCK_SET_CODE);
        expect(repoSaveCards).toHaveBeenCalledWith(mockSetCards);
        expect(foundCard).toEqual(mockSavedCard);
    });

    it('returns unique instance of a card by uuid', async () => {
        const repoFindByUuid = jest.spyOn(repository, 'findByUuid');
        const foundCard: CardDto | null = await service.findByUuid(inputCard.uuid);
        expect(repoFindByUuid).toHaveBeenCalledWith(inputCard.uuid);
        expect(foundCard).toEqual(mockSavedCard);
    });
});
