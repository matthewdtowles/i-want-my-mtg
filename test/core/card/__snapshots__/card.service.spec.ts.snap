// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardService should find all cards in a set 1`] = `
[
  {
    "artist": "artist",
    "id": 1,
    "imgSrc": "https://cards.scryfall.io/small/front/1/a/1abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 1,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "1",
      "w",
    ],
    "name": "Test Card Name 1",
    "number": "1",
    "oracleText": "Test card text.",
    "prices": [
      {
        "cardId": 1,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "10.00",
        "normal": "5.00",
      },
      {
        "cardId": 1,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "11.00",
        "normal": "6.00",
      },
      {
        "cardId": 1,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "12.00",
        "normal": "7.00",
      },
    ],
    "rarity": "common",
    "set": {
      "baseSize": 3,
      "cards": [],
      "code": "SET",
      "keyruneCode": "set",
      "name": "Test Set",
      "releaseDate": "2022-01-01",
      "type": "expansion",
      "url": "/set/set",
    },
    "setCode": "SET",
    "type": undefined,
    "url": "/card/set/1",
    "uuid": undefined,
  },
  {
    "artist": "artist",
    "id": 2,
    "imgSrc": "https://cards.scryfall.io/small/front/2/a/2abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 2,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "2",
      "w",
    ],
    "name": "Test Card Name 2",
    "number": "2",
    "oracleText": "Test card text.",
    "prices": [
      {
        "cardId": 2,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "10.00",
        "normal": "5.00",
      },
      {
        "cardId": 2,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "11.00",
        "normal": "6.00",
      },
      {
        "cardId": 2,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "12.00",
        "normal": "7.00",
      },
    ],
    "rarity": "uncommon",
    "set": {
      "baseSize": 3,
      "cards": [],
      "code": "SET",
      "keyruneCode": "set",
      "name": "Test Set",
      "releaseDate": "2022-01-01",
      "type": "expansion",
      "url": "/set/set",
    },
    "setCode": "SET",
    "type": undefined,
    "url": "/card/set/2",
    "uuid": undefined,
  },
  {
    "artist": "artist",
    "id": 3,
    "imgSrc": "https://cards.scryfall.io/small/front/3/a/3abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 3,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "3",
      "w",
    ],
    "name": "Test Card Name 3",
    "number": "3",
    "oracleText": "Test card text.",
    "prices": [
      {
        "cardId": 3,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "10.00",
        "normal": "5.00",
      },
      {
        "cardId": 3,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "11.00",
        "normal": "6.00",
      },
      {
        "cardId": 3,
        "date": 2022-01-01T00:00:00.000Z,
        "foil": "12.00",
        "normal": "7.00",
      },
    ],
    "rarity": "common",
    "set": {
      "baseSize": 3,
      "cards": [],
      "code": "SET",
      "keyruneCode": "set",
      "name": "Test Set",
      "releaseDate": "2022-01-01",
      "type": "expansion",
      "url": "/set/set",
    },
    "setCode": "SET",
    "type": undefined,
    "url": "/card/set/3",
    "uuid": undefined,
  },
]
`;

exports[`CardService should save legalities and delete previously saved legalities not given 1`] = `
[
  {
    "artist": undefined,
    "id": 1,
    "imgSrc": "https://cards.scryfall.io/small/front/imgSrc",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 1,
        "format": "legacy",
        "status": "legal",
      },
    ],
    "manaCost": null,
    "name": "Card 1",
    "number": "1",
    "oracleText": undefined,
    "prices": [],
    "rarity": "common",
    "set": null,
    "setCode": "SET",
    "type": "type",
    "url": "/card/set/1",
    "uuid": "uuid-123",
  },
]
`;

exports[`CardService should save new cards and return saved cards 1`] = `
[
  {
    "artist": "artist",
    "id": 1,
    "imgSrc": "https://cards.scryfall.io/small/front/1/a/1abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 1,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "1",
      "w",
    ],
    "name": "Test Card Name 1",
    "number": "1",
    "oracleText": "Test card text.",
    "prices": [],
    "rarity": "common",
    "set": null,
    "setCode": "SET",
    "type": "type",
    "url": "/card/set/1",
    "uuid": "abcd-1234-efgh-5678-ijkl-SET1",
  },
  {
    "artist": "artist",
    "id": 2,
    "imgSrc": "https://cards.scryfall.io/small/front/2/a/2abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 2,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "2",
      "w",
    ],
    "name": "Test Card Name 2",
    "number": "2",
    "oracleText": "Test card text.",
    "prices": [],
    "rarity": "uncommon",
    "set": null,
    "setCode": "SET",
    "type": "type",
    "url": "/card/set/2",
    "uuid": "abcd-1234-efgh-5678-ijkl-SET2",
  },
  {
    "artist": "artist",
    "id": 3,
    "imgSrc": "https://cards.scryfall.io/small/front/3/a/3abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 3,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "3",
      "w",
    ],
    "name": "Test Card Name 3",
    "number": "3",
    "oracleText": "Test card text.",
    "prices": [],
    "rarity": "common",
    "set": null,
    "setCode": "SET",
    "type": "type",
    "url": "/card/set/3",
    "uuid": "abcd-1234-efgh-5678-ijkl-SET3",
  },
]
`;

exports[`CardService should update existing cards and return saved cards 1`] = `
[
  {
    "artist": "artist",
    "id": 1,
    "imgSrc": "https://cards.scryfall.io/small/front/1/a/1abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 1,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 1,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "1",
      "w",
    ],
    "name": "Test Card Name 1",
    "number": "1",
    "oracleText": "Test card text.",
    "prices": [],
    "rarity": "common",
    "set": null,
    "setCode": "SET",
    "type": "type",
    "url": "/card/set/1",
    "uuid": "abcd-1234-efgh-5678-ijkl-SET1",
  },
  {
    "artist": "artist",
    "id": 2,
    "imgSrc": "https://cards.scryfall.io/small/front/2/a/2abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 2,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 2,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "2",
      "w",
    ],
    "name": "Test Card Name 2",
    "number": "2",
    "oracleText": "Test card text.",
    "prices": [],
    "rarity": "uncommon",
    "set": null,
    "setCode": "SET",
    "type": "type",
    "url": "/card/set/2",
    "uuid": "abcd-1234-efgh-5678-ijkl-SET2",
  },
  {
    "artist": "artist",
    "id": 3,
    "imgSrc": "https://cards.scryfall.io/small/front/3/a/3abc123def456.jpg",
    "isReserved": false,
    "legalities": [
      {
        "cardId": 3,
        "format": "standard",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "commander",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "modern",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "legacy",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "vintage",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "brawl",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "explorer",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "historic",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "oathbreaker",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "pauper",
        "status": "legal",
      },
      {
        "cardId": 3,
        "format": "pioneer",
        "status": "legal",
      },
    ],
    "manaCost": [
      "3",
      "w",
    ],
    "name": "Test Card Name 3",
    "number": "3",
    "oracleText": "Test card text.",
    "prices": [],
    "rarity": "common",
    "set": null,
    "setCode": "SET",
    "type": "type",
    "url": "/card/set/3",
    "uuid": "abcd-1234-efgh-5678-ijkl-SET3",
  },
]
`;
